### PR TITLE
Drag & drop to rearrange question

### PR DESCRIPTION
**Pull request by govind007**
Screenshot 2023-10-02 at 11 14 15 AM
Screenshot 2023-10-02 at 11 14 21 AM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,8 +32,11 @@
     "react-dom": "^17.0.2",
     "react-helmet": "^6.1.0",
     "react-hook-form": "^7.27.1",
+    "react-router": "^6.16.0",
```
Comment (rajatkantinandi) [inline package.json:35]: Strange, did we not have this before? How was routing working before?

File: src/components/QuestionView/Collapsed.tsx [hunk 2]
```diff
@@ -19,18 +19,12 @@ export default function CollapsedView({
...
 
   return (
-    <Card
-      shadow="sm"
-      p="lg"
-      my="sm"
-      withBorder
-      className="secondaryCard clickable slideUp"
-      onClick={() => setExpandedQuestionIndex(questionNum - 1)}>
+    <Card shadow="sm" p="lg" my="sm" withBorder className="secondaryCard clickable slideUp">
```
Comment (rajatkantinandi) [inline src/components/QuestionView/Collapsed.tsx]: We were able to expand by clicking on the card instead of just the caret icon. Why are we changing this behaviour?

File: src/components/QuestionView/Collapsed.tsx [hunk 3]
```diff
@@ -54,9 +48,12 @@ export default function CollapsedView({
           <ActionIcon variant="transparent" title="Edit" onClick={setActiveQuestion}>
             <Icon name="pencil" width={22} />
           </ActionIcon>
-          <ActionIcon variant="transparent">
+          <ActionIcon variant="transparent" onClick={() => setExpandedQuestionId(question.questionId)}>
             <Icon name="caretDown" />
           </ActionIcon>
+          <ActionIcon variant="transparent" className="questionHandle">
```
Comment (rajatkantinandi) [inline src/components/QuestionView/Collapsed.tsx:53]: Does it show up always? Can you add screenshots in the PR description?

File: src/components/QuestionView/Expanded.tsx [hunk 4]
```diff
@@ -31,8 +31,8 @@ export default function ExpandedView({
       : {};
 
   return (
-    <Card shadow="sm" p="lg" my="sm" withBorder className="secondaryCard slideDown">
-      <Card className="secondaryCard clickable" p={0} onClick={() => setExpandedQuestionIndex(null)}>
```
Comment (rajatkantinandi) [inline src/components/QuestionView/Expanded.tsx]: Same question. We should be able collapse by clicking on the header like today.

File: src/components/QuestionView/Expanded.tsx [hunk 5]
```diff
@@ -50,12 +50,15 @@ export default function ExpandedView({
             <ActionIcon variant="transparent" title="Edit" onClick={setActiveQuestion}>
               <Icon name="pencil" width={22} />
             </ActionIcon>
-            <ActionIcon variant="transparent">
+            <ActionIcon variant="transparent" onClick={() => setExpandedQuestionId(null)}>
               <Icon name="caretUp" />
             </ActionIcon>
+            <ActionIcon variant="transparent" className="questionHandle">
```
Comment (rajatkantinandi) [inline src/components/QuestionView/Expanded.tsx]: Do we need to drag in expanded mode? Dragging in expanded mode will be difficult.
Maybe we show it in collapsed mode only. And only show it after the user clicks the rearrange questions button and then we collapse all the questions. Does not make sense to show the drag icon all the time. Users might accidentally drag stuff.

File: src/routes/ConfigureQuiz/QuestionCard.tsx [hunk 6]
```diff
@@ -93,43 +94,73 @@ export default function QuestionCard({
...
+  function onQuestionSwap(data) {
+    const updatedQuestionIds = data.map((x) => x.questionId).join(',');
+    const initialQuestionIds = questions.map((x) => x.questionId).join(',');
+
+    if (initialQuestionIds !== updatedQuestionIds) {
+      replace(
+        data.map((x) => {
+          delete x.id;
+          return x;
+        }),
```
Comment (rajatkantinandi) [inline src/routes/ConfigureQuiz/QuestionCard.tsx]: Did not understand this part. What is going on here?

File: src/routes/ConfigureQuiz/QuestionCard.tsx [hunk 7]
```diff
@@ -93,43 +94,73 @@ export default function QuestionCard({
...
-            setExpandedQuestionIndex={setExpandedQuestionIndex}
-          />
-        ),
-      )}
+      <ReactSortable
+        list={questions.map((item, idx) => ({ ...item, id: idx + 1, name: item.text }))}
+        chosenClass={styles.chosenStyle}
+        handle=".questionHandle"
+        onChoose={(data, ev: any) => {
+          if (!!activeQuestionIndex) {
```
Comment (rajatkantinandi) [inline src/routes/ConfigureQuiz/QuestionCard.tsx]: We won't need this if we don't allow dragging in expanded questions.

File: src/components/QuestionEdit/index.tsx [hunk 8]
```diff
@@ -144,7 +144,7 @@ export default function QuestionEdit({
 
   function onDeleteClick(ev) {
     ev.preventDefault();
-    deleteQuestion();
+    deleteQuestion(ev);
```
Comment (rajatkantinandi) [inline src/components/QuestionEdit/index.tsx:147]: Why this change?

**Review by rajatkantinandi — CHANGES_REQUESTED** [review]
My suggestion is to not show the drag icon always and only show it when users click a button at the top of the question list and only drag collapsed questions (We can disable the button when editing a question). Otherwise, it will be messy, too many icons in the UI and people will accidentally click those icons.

When showing drag icon (drag mode on), we can hide delete, edit buttons.

File: src/styles/variables.css [hunk 9]
```diff
@@ -42,6 +42,7 @@
   --quiz-card-bg-3: #e91e63;
   --quiz-card-bg-4: #4caf50;
   --quiz-card-bg-5: #673ab7;
+  --dragging-question-bg: #d7e7e9
```
Comment (rajatkantinandi) [inline src/styles/variables.css]:
```bash
Syntax error: Missed semicolon (47:3)
```

File: package.json (hunk 1 above)
Comment (govind007) [reply package.json:35]: it was not in package.json. might be in node_modules as a dependency of another package

File: src/components/QuestionEdit/index.tsx (hunk 8 above)
Comment (govind007) [reply src/components/QuestionEdit/index.tsx:147]: we are using ev.stopPropagation in deleteQuestion function. it was throwing console error

File: src/routes/ConfigureQuiz/QuestionCard.tsx (hunk 7 above)
Comment (govind007) [reply src/routes/ConfigureQuiz/QuestionCard.tsx]: fixed

File: src/routes/ConfigureQuiz/QuestionCard.tsx (hunk 6 above)
Comment (govind007) [reply src/routes/ConfigureQuiz/QuestionCard.tsx]: for ReactSortable, I need a `id` for each question, So in below code I'm adding `id` key in ReactSortable list data. and here I'm removing `id` key before updating questions data

File: src/components/QuestionView/Expanded.tsx (hunk 5 above)
Comment (govind007) [reply src/components/QuestionView/Expanded.tsx]: ok

File: src/components/QuestionView/Collapsed.tsx (hunk 3 above)
Comment (govind007) [reply src/components/QuestionView/Collapsed.tsx:53]: fixed

File: src/components/QuestionView/Expanded.tsx (hunk 4 above)
Comment (govind007) [reply src/components/QuestionView/Expanded.tsx]: fixed

File: src/components/QuestionView/Collapsed.tsx (hunk 2 above)
Comment (govind007) [reply src/components/QuestionView/Collapsed.tsx]: fixed

File: src/routes/ConfigureQuiz/QuestionCard.tsx (hunk 6 above)
Comment (rajatkantinandi) [reply src/routes/ConfigureQuiz/QuestionCard.tsx]: Adding comments

**Review by rajatkantinandi — APPROVED** [review]
